### PR TITLE
[RPMS] Enable CPU and Memory accounting during node installation

### PR DESCRIPTION
**Pull request by sdodson**
Related to https://github.com/kubernetes/kubernetes/pull/17688#issuecomment-220447268

### CODE REVIEW

**Comment by sdodson** [comment]
@derekwaynecarr PTAL

**Comment by sdodson** [comment]
@tdawson mind reviewing the specfile changes?

File: origin.spec [hunk 1]
```diff
@@ -413,6 +418,10 @@ fi
 
 %post node
 %systemd_post %{name}-node.service
+# If accounting is not currently enabled systemd reexec
+if [[ `systemctl show mananager | grep -q -e CPUAccounting=no -e MemoryAccounting=no` == 0 ]]; then
```
Comment (ncdc) [inline origin.spec]: `mananager`?

Comment (derekwaynecarr) [inline origin.spec]: typo?

Comment (sdodson) [inline origin.spec]: Yeah, looks look `systemctl show bogus` works too though i want to verify that's not just unexpected behavior.

Comment (sdodson) [inline origin.spec]: I changed it to check for docker and our node service to avoid possible collisions with another service that may have explicitly disabled accounting. I've not found a better way to figure out the real default.

**Comment by smarterclayton** [comment]
@csrwng something we should keep in mind with `cluster up`

**Comment by derekwaynecarr** [comment]
The change looks good to me.

@jeremyeder - the moment we launched a docker container, it would turn accounting on for its parent slices.  So for example, `docker-xyz.scope` in `system.slice` would enable accounting, and then by default `system.slice` would enable accounting and `-.slice` would enable accounting after the fact.  We need accounting on by default so `docker.service` is accounted, `kubelet/node` agent is accounted, and that we have accurate roll-up to the top-level slices reported in cadvisor.

https://www.freedesktop.org/software/systemd/man/systemd.resource-control.html#CPUAccounting=

File: origin.spec (hunk 1 above)
Comment (derekwaynecarr) [inline origin.spec]: You could do `systemctl show systemd`

**Comment by jeremyeder** [comment]
We have no other users of this flag that I'm aware of...asking around internally and cc'd you.  To mock this I believe I can use OSE3.2 and enable accounting in the atomic-openshift-node.service unitfile:
```
CPUAccounting=on
MemoryAccounting=on
```

What about BlockIOAccounting?

File: origin.spec (hunk 1 above)
Comment (sdodson) [inline origin.spec]: This is the same as `mananager` basically it's returning the defaults because it's a service that doesn't exist. According to folks #systemd it's not possible to get the true default until v228 where you'd use `systemctl show --global` I don't have a host running v228 or newer so I've not verified that. In absence of a true method to return the default I'd prefer checking the two services we really care about.

Comment (sdodson) [inline origin.spec]: RHEL 7.2 has systemd v219.

**Comment by sdodson** [comment]
@jeremyeder Derek decided we'd prefer having accounting enabled globally rather than for just our two services.

**Comment by sdodson** [comment]
@derekwaynecarr and @jeremyeder Are you two ok with us merging this?

**Comment by jeremyeder** [comment]
Situation as I see it:

1) As far as I know, no RH product has ever used these.  
2) If we want to ever have proper systemd support then we really have no choice.
3) Once this is on, there's really no going back.  Any issues will have to be resolved.
4) We have no idea what the overheads or other impacts might be.

**Comment by derekwaynecarr** [comment]
Situation as I see it:

Kubernetes not on systemd
* requires CPU and memory accounting globally
* parents existing system pids in /system
* parents kubelet in /kubelet
* parents docker in /docker

Kubernetes on systemd
* should require CPU and memory accounting globally
* should not move existing pids, but needs full accounting on /system.slice
* should get accounting specific to kubelet/node (will get via unit having this global default)
* should launch containers end-user pods under machine.slice (coming soon)

So basically, nodes need accounting, and accurate accounting to ever truly support utilization aware scheduling.

**Comment by derekwaynecarr** [comment]
so to summarize: this should be merged.

**Comment by derekwaynecarr** [comment]
@jeremyeder - we do not need BlkIO today.

**Comment by derekwaynecarr** [comment]
In addition, the moment we launch a container by default today:

1. cpu and memory accounting is enabled in /system.slice/docker-xyz
2. cpu and memory accounting is enabled in /system.slice (but is no longer accurate since it only catches new pids)
3. cpu and memory accounting is enabled in -.slice (but is no longer accurate since it catches new pids)

This fix just ensures its accurate from the start.

see: https://www.freedesktop.org/software/systemd/man/systemd.resource-control.html#CPUAccounting=

"Note that turning on CPU accounting for one unit will also implicitly turn it on for all units contained in the same slice and for all its parent slices and the units contained therein."

@jeremyeder - agree with above?

**Comment by derekwaynecarr** [comment]
and by "no longer accurate" - previously launched pids are charged to the "/" cgroup rather than the cgroup returned in systemd-cgls, where as enabling this from the start ensures they match (as expected)

**Comment by sdodson** [comment]
@derekwaynecarr What I've implemented enables it globally not just for docker and the node. That's still what you want, correct?

**Comment by derekwaynecarr** [comment]
@sdodson - yes, i want it globally.

**Comment by mdshuai** [comment]
@derekwaynecarr @sdodson
I's very curios about our current ose env, Both on master and node "DefaultCPUAccounting" and "CPUAccounting" is not enabled. But the cpu and memroy cgroup has different value.
On node the cpuacct,cpu and memory is /system.slice/docker.service
But on master  the cpuacct,cpu and memory is /
##### master:
```
[root@openshift-149 ~]# cat /proc/$(pidof docker-current)/cgroup
10:perf_event:/
9:cpuacct,cpu:/
8:freezer:/
7:memory:/
6:cpuset:/
5:blkio:/
4:net_cls:/
3:devices:/system.slice/docker.service
2:hugetlb:/
1:name=systemd:/system.slice/docker.service
```
#####  node:
```
[root@openshift-150 ~]# cat /proc/$(pidof docker-current)/cgroup
10:hugetlb:/
9:net_cls:/
8:memory:/system.slice/docker.service
7:freezer:/
6:cpuacct,cpu:/system.slice/docker.service
5:devices:/system.slice/docker.service
4:cpuset:/
3:blkio:/system.slice/docker.service
2:perf_event:/
1:name=systemd:/system.slice/docker.service

```
This env setup by openshift-ansible. question: 
1) what's different cause this different ?

**Comment by derekwaynecarr** [comment]
@mdshuai - if you restart docker on the master, do you get a different result?  if you restart the master after install, do you get a different result?

**Comment by mdshuai** [comment]
@derekwaynecarr After reboot master/node instance same result as above. master: http://pastebin.test.redhat.com/380179  node: http://pastebin.test.redhat.com/380180

If I manually enable  "DefaultCPUAccounting" and "DefaultMemoryAccounting" on master then reboot it, it can become as what you said. I don't know why on node I don't enable cpu/memory accounting, it still can be memory:/system.slice/docker.service
Any other config can I check ?

**Comment by derekwaynecarr** [comment]
@sdodson - didn't this PR just change the node and not the master?  

@mdshuai - changing global systemd defaults will only take effect either a) after a reboot, or b) after a systemctl daemon reload and a restart of services that you want to pick up new global defaults.  In your example, it should be picking up the values in `/etc/systemd/system.conf.d/origin-accounting.conf`.

**Comment by sdodson** [comment]
> @sdodson - didn't this PR just change the node and not the master?

It changes the system wide default and reloads systemd. So it'd affect all processes.

**Comment by mdshuai** [comment]
@derekwaynecarr In my env there is no ```/etc/systemd/system.conf.d/origin-accounting.conf``` I don't set it.

**Comment by sdodson** [comment]
@mdshuai This PR is an RPM specfile change. When applied and a user installs the node RPM it will drop the file in /etc/systemd/system.conf.d/origin-accounting.conf and then run `systemctl daemon-reexec` can you manually do those steps and see what the outcome is?

**Comment by sdodson** [comment]
@derekwaynecarr should we merge this or not?

File: origin.spec [hunk 2]
```diff
@@ -413,6 +418,10 @@ fi
 
 %post node
 %systemd_post %{name}-node.service
+# If accounting is not currently enabled systemd reexec
+if [[ `systemctl show docker %{name}-node | grep -q -e CPUAccounting=no -e MemoryAccounting=no` == 0 ]]; then
```
Comment (ingvagabund) [inline origin.spec]:
```bash
`systemctl show docker %{name}-node | grep -q -e CPUAccounting=no -e MemoryAccounting=no; echo $?` == 0
```

Comment (sdodson) [inline origin.spec]: Would this be best?
```
[[ `systemctl show docker %{name}-node | grep -q -e CPUAccounting=no -e MemoryAccounting=no` -eq 0 ]]
```

Comment (ingvagabund) [inline origin.spec]: In general, `command` does not return return code. `systemctl show docker %{name}-node | grep -q -e CPUAccounting=no -e MemoryAccounting=no` always returns an empty string. Which is then (most likely) converted to 0. So the condition will be always true. The only I am aware to test for RC is to check ``$?``.

**Comment by derekwaynecarr** [comment]
[merge]  - accounting needs to be enabled globally.
